opm/output/data/Wells.hpp

```diff
@@ -316,16 +319,149 @@ namespace Opm {
         }
     };
 
-    struct Segment {
-        Rates rates;
-        SegmentPressures pressures;
-        std::size_t segNumber;
+    class SegmentPhaseQuantity
+    {
+    public:
+        enum class Item : std::size_t {
+            Oil, Gas, Water,
+
+            // -- Must be last enumerator --
+            NumItems,
+        };
+
+        void clear()
+        {
+            this->mask_.reset();
+            this->value_.fill(0.0);
+        }
+
+        constexpr bool has(const Item p) const
+        {
+            const auto i = this->index(p);
+
+            return (i < Size) && this->mask_[i];
+        }
+
+        bool operator==(const SegmentPhaseQuantity& vec) const
+        {
+            return (this->mask_  == vec.mask_)
+                && (this->value_ == vec.value_);
+        }
+
+        double get(const Item p) const
+        {
+            if (! this->has(p)) {
+                throw std::invalid_argument {
+                    "Request for Unset Item Value for " + this->itemName(p)
+                };
+            }
+
+            return this->value_[ this->index(p) ];
+        }
+
+        SegmentPhaseQuantity& set(const Item p, const double value)
+        {
+            const auto i = this->index(p);
+
+            if (i >= Size) {
+                throw std::invalid_argument {
+                    "Cannot Assign Item Value for Unsupported Item '"
+                    + this->itemName(p) + '\''
+                };
+            }
+
+            this->mask_.set(i);
+            this->value_[i] = value;
+
+            return *this;
+        }
+
+        template <class MessageBufferType>
+        void write(MessageBufferType& buffer) const
+        {
+            auto maskrep = this->mask_.to_ullong();
+            buffer.write(maskrep);
+
+            for (const auto& x : this->value_) {
+                buffer.write(x);
+            }
+        }
+
+        template <class MessageBufferType>
+        void read(MessageBufferType& buffer)
+        {
+            this->clear();
+
+            {
+                auto mask = 0ull;
+                buffer.read(mask);
+
+                this->mask_ = std::bitset<Size>(mask);
+            }
+
+            for (auto& x : this->value_) {
+                buffer.read(x);
+            }
+        }
+
+        template <class Serializer>
+        void serializeOp(Serializer& serializer)
+        {
+           serializer(this->mask_);
+           serializer(this->value_);
+        }
+
+        static SegmentPhaseQuantity serializationTestObject()
+        {
+            return SegmentPhaseQuantity{}
+                .set(Item::Oil  ,  1.0)
+                .set(Item::Gas  ,  7.0)
+                .set(Item::Water, 29.0);
+        }
+
+    private:
+        enum { Size = static_cast<std::size_t>(Item::NumItems) };
+
+        std::bitset<Size>        mask_{};
+        std::array<double, Size> value_{};
```

Comment:
Some documentation would not hurt. After reading the whole file, I understand that mask_ is used to save whether a value is set or not. Not clear without further reading.

I guess `std::bitset` instead of storing std::optional is used to save space. I there a case where the initial values 0 are important or could one use `std::numeric_limits<double>::max()` to mark unset Items?

Reply:
> understand that mask_ is used to save whether a value is set or not. Not clear without further reading.

I've renamed that object to `has_` and added a comment.

> I guess `std::bitset` instead of storing std::optional is used to save space.

It's actually more because it's easier to handle `std::bitset` than `std::optional` in the [`read()`](https://github.com/OPM/opm-common/blob/19babe34a18d9b730ca48784dad417e676b54fb5/opm/output/data/Wells.hpp#L391) and [`write()`](https://github.com/OPM/opm-common/blob/19babe34a18d9b730ca48784dad417e676b54fb5/opm/output/data/Wells.hpp#L380) operations invoked as part of collecting the results onto the I/O rank.  In the case of libstdc++ `std::bitset<N>` is ultimately backed by a built-in array of `unsigned long`, with the number of array elements adjusted to get at least `N` bits.  On a 64-bit platform this **typically** means that there's a single array elements for bitsets of sizes 1 through 64.

> Is there a case where the initial values `0` are important

We need to be able to have `0` as a valid value, especially for the "holdup" fractions.

> could one use `std::numeric_limits<double>::max()` to mark unset Items?

I suppose it's possible, but I would prefer not to make decisions based on `x == max()` (or `x != max()` or `x < max()` for that matter).

Reply:
> I suppose it's possible, but I would prefer not to make decisions based on `x == max()` (or `x != max()` or `x < max()` for that matter).

Well, of course you would hide that in your `has(...)` methods or whatever they are called. 
My main concerns are mainly memory consumption both for storing and communicating, but course there are a lot less segments than cells.

Reply:
>>  would prefer not to make decisions based on `x == max()` (or `x != max()` or `x < max(`) for that matter).
>
> Well, of course you would hide that in your `has(...)` methods or whatever they are called.

True, but I still don't want to make decisions based on special values that may or may not go through MPI unchanged.